### PR TITLE
Fix excessive memory usage in aggregator

### PR DESCRIPTION
**Pull request by Modularius**
## Summary of changes

- Replaces `JoinSet` method of producing `frame_event_list` messages to Kakfa, with a dedicated long-running thread which is tasked with emitting frames to Kafka.
- Every call to `cache_poll` drains the cache of completed frames from the front, until an incomplete frame is reached, and sends them to the dedicated thread.
- Frames are emitted in the order they were begun (i.e. when its first digitiser event message arrived).

## Instruction for review/testing

General code review. This change been tested and verified with simulated data, and on HiFi.

Closes https://github.com/STFC-ICD-Research-and-Design/supermusr-data-pipeline/issues/265.

### CODE REVIEW

File: digitiser-aggregator/src/main.rs [hunk 1]
```diff
@@ -81,6 +88,8 @@ struct Cli {
     otel_level: LevelFilter,
 }
 
+type ChannelSender = Sender<AggregatedFrame<EventData>>;
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:91]: Could this type have a more descriptive name? `ChannelSender` is rather abstract.

File: digitiser-aggregator/src/main.rs [hunk 2]
```diff
@@ -139,23 +148,30 @@ async fn main() -> anyhow::Result<()> {
         "Number of complete frames sent by the aggregator"
     );
 
-    let mut kafka_producer_thread_set = JoinSet::new();
-
     let mut cache_poll_interval = tokio::time::interval(Duration::from_millis(args.cache_poll_ms));
+
+    // Creates Send-Frame thread and returns channel sender
+    let channel_send = create_producer_thread(
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:154]:
```suggestion
    let channel_send = create_producer_task(
```

File: digitiser-aggregator/src/main.rs [hunk 3]
```diff
@@ -172,12 +188,10 @@ fn spanned_root_as_digitizer_event_list_message(
...
 async fn process_kafka_message(
     use_otel: bool,
-    kafka_producer_thread_set: &mut JoinSet<()>,
+    channel_send: &ChannelSender,
     cache: &mut FrameCache<EventData>,
-    producer: &FutureProducer,
-    output_topic: &str,
     msg: &BorrowedMessage<'_>,
-) {
+) -> anyhow::Result<()> {
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:194]: Could something other than `anyhow::Result` be used here?

File: digitiser-aggregator/src/main.rs [hunk 4]
```diff
@@ -232,13 +239,10 @@ async fn process_kafka_message(
...
 async fn process_digitiser_event_list_message(
-    use_otel: bool,
-    kafka_producer_thread_set: &mut JoinSet<()>,
+    channel_send: &ChannelSender,
     cache: &mut FrameCache<EventData>,
-    producer: &FutureProducer,
-    output_topic: &str,
     msg: DigitizerEventListMessage<'_>,
-) {
+) -> anyhow::Result<()> {
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:245]: Same here.

File: digitiser-aggregator/src/main.rs [hunk 5]
```diff
@@ -266,51 +263,103 @@ async fn process_digitiser_event_list_message(
...
+    Ok(())
 }
 
 #[tracing::instrument(skip_all, level = "trace")]
 async fn cache_poll(
-    use_otel: bool,
-    kafka_producer_thread_set: &mut JoinSet<()>,
+    channel_send: &ChannelSender,
     cache: &mut FrameCache<EventData>,
+) -> anyhow::Result<()> {
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:273]: Same here.

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: digitiser-aggregator/src/main.rs [hunk 6]
```diff
@@ -81,6 +88,9 @@ struct Cli {
     otel_level: LevelFilter,
 }
 
+type AggregatedFrameToBufferSender = Sender<AggregatedFrame<EventData>>;
+type TrySendAggregatedFrameError = TrySendError<AggregatedFrame<EventData>>;
```
[the remark this answers is not in the document]
Comment (Modularius) [reply digitiser-aggregator/src/main.rs:92]: Unfortunately, Rust won't let you implement `Error` for `()`.

**Review by DanNixon — APPROVED** [review]

**Pull request merged**